### ISSUE TITLE
Server misbehaves when no users are connected

### ISSUE THREAD

**Issue by gwillen** [gwillen/solstice-audio-test#104]
Since the addition of the "server-to-server" API, it is possible (and in fact likely) that the backing track will be started slightly _before_ any users connect to listen to it. This is a problem because right now the bucket brigade server clears all its internal state on every request if there are no users.

In `clean_users`:
```
    if not users:
        state.reset()
```

I'm not sure what the best approach to this would be. Some options:
- Never reset server state. The server should work fine without this. (I assume this is undesirable because the server has a bunch of options now, like the metronome, which it's nice to clear between customers.)
- Never reset server state if any users are around OR there is a backing track active. This seems like it could work. It will delay server state reset by at most a minute or two while the backing track finishes, if everybody leaves. (In fact, if someone is testing/singing by themselves, this will protect them from the server blowing itself up if their connection is interrupted for 5 seconds, which is probably desirable. Or if the server's connection is interrupted for 5 seconds, for that matter.)
- Make a solstice branch, and never reset server state on the solstice branch. Leave the master branch alone. (This is probably annoying to deal with.)
 

There's an important caveat here, which is that I think fixing this will expose another issue related to backing track timing and last_request_clock. Right now, if the server doesn't get any requests for a long time, last_request_clock can get very stale, which would cause problems with the buffer-clearing logic; and the backing-track logic, being part of the buffer-clearing logic, actually "jumps the gun" by starting the backing track at last_request_clock instead of server_clock. Under normal circumstances this will be significantly less than a second, and right now the state reset logic will reset it to None if it gets more than 5 seconds stale. Without that, it could be many seconds old.

Two approaches I could see using here:
- Fix the song-start logic to start the song at server_clock instead of last_request clock. Cap last_request_clock at server_clock - QUEUE_LENGTH to prevent other problems.
- Hack it a bit, by adding an artificial "request" every 500ms or so to keep things moving.

**Comment by gwillen** [comment]
@jeffkaufman should probably opine here.

**Comment by jeffkaufman** [comment]
What would you think of never resetting the server state if we have ever received a server to server response? I think it is reasonable to say that either:

* There is no managing server, in which case the bucket brigade server should do its best

* There is a managing server, in which case the bucket brigade server should fully delegate state management to the other server

As for last request clock going stale, I agree that's kind of awkward; bucket brigade wasn't designed for this case where there are no users connected. What would you think of having the server to server request update the last request clock? I think if you do that then everything just works?

**Comment by dspeyer** [comment]
This seems like it breaks ritualEngine a lot.  No users connected while the server-to-server api sets backing track is the default use case.

I think just turning this off for solstice should work.  Maybe there should be a reset_everything api?  Or maybe stop_singing should do that?  If we need to have a turn_off_auto_reset api, ritualEngine could easily hit that at startup and/or before setting backing track.

**Comment by gwillen** [comment]
I like Jeff's suggestion and will implement that -- if we get a server-to-server API call, we will disable the automatic resetting. (Forever, until we are restarted.)

I will also make server-to-server requests update last_request_clock (which means they will also need to perform buffer-clearing and backing-track processing, but that's fine.) I think that will require some refactoring, and almost but not quite fixes the problem in theory -- I think it would still be the case that if you start a song after a long pause with no requests, it will start at the wrong time. However I think this will actually not affect Ritual Engine as-implemented now, since it always makes two requests back-to-back (mark_stop_singing followed by mark_start_singing). I don't want to count on that but maybe it means I can punt on the proper fix to that issue for now.

**Comment by dspeyer** [comment]
Just in case it's relevant, I should clarify that ritual engine's current
behavior is to send stop_singing *both* at the end of a song *and* right
before starting one.  Defense in depth.

**Comment by gwillen** [comment]
That seems great, and should be harmless.

**Comment by gwillen** [comment]
I have the fix for this on gwillen-dev and am testing it.

**Comment by gwillen** [comment]
Fixed in https://github.com/gwillen/solstice-audio-test/commit/292131c9feb7df3219e664598de31e4afe08a59b and tested.
